appcues/src/main/java/com/appcues/ui/ExperienceRenderer.kt

```diff
@@ -11,52 +12,40 @@ import com.appcues.data.model.ExperiencePriority.NORMAL
 import com.appcues.data.model.ExperienceTrigger
 import com.appcues.data.model.RenderContext
 import com.appcues.statemachine.Action.EndExperience
-import com.appcues.statemachine.Action.StartExperience
 import com.appcues.statemachine.Action.StartStep
 import com.appcues.statemachine.Error
-import com.appcues.statemachine.Error.NoActiveStateMachine
 import com.appcues.statemachine.State
 import com.appcues.statemachine.State.Idling
 import com.appcues.statemachine.StateMachine
-import com.appcues.statemachine.StateMachine.StateMachineListener
 import com.appcues.statemachine.StateMachineFactory
 import com.appcues.statemachine.StepReference
 import com.appcues.util.ResultOf
 import com.appcues.util.ResultOf.Failure
 import com.appcues.util.ResultOf.Success
+import kotlinx.coroutines.Dispatchers
 import kotlinx.coroutines.flow.MutableSharedFlow
-import kotlinx.coroutines.flow.SharedFlow
+import kotlinx.coroutines.flow.collectLatest
+import kotlinx.coroutines.flow.filter
+import kotlinx.coroutines.launch
 import kotlin.collections.set
 
 internal class ExperienceRenderer(
     private val config: AppcuesConfig,
+    private val appcuesCoroutineScope: AppcuesCoroutineScope,
     private val experienceLifecycleTracker: ExperienceLifecycleTracker,
     private val repository: AppcuesRepository,
     private val sessionMonitor: SessionMonitor,
     private val stateMachineFactory: StateMachineFactory,
-) : StateMachineListener {
+) {
 
     private val stateMachines: HashMap<RenderContext, StateMachine> = hashMapOf()
 
-    private val _stateFlow = MutableSharedFlow<State>(1)
-    val stateFlow: SharedFlow<State>
-        get() = _stateFlow
+    private val _allStates = MutableSharedFlow<State>(1)
 
-    private val _errorFlow = MutableSharedFlow<Error>(1)
-    val errorFlow: SharedFlow<Error>
-        get() = _errorFlow
+    suspend fun collectState(renderContext: RenderContext, action: suspend (value: State) -> Unit) =
+        _allStates.filter { it.experience.renderContext == renderContext }.collectLatest(action)
```

Comment:
is it necessary to have this `_allStates` aggregate flow? Could the VM usage instead just listen to the `StateMachine` for it's `RenderContext` directly?

Reply:
probably, I was imagining we would want allStates for debugger later, but maybe not